File: api/Enkl.Api/Controllers/TelemetryController.cs
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;

namespace Enkl.Api.Controllers;

/// <summary>
/// Anonymous Real User Monitoring beacon — no Authorize policy at all (unlike every other
/// controller in this API), since it's a fire-and-forget report from every page load, signed in or
/// not. See Program.cs's MustChangePassword-enforcement middleware, which explicitly excludes this
/// route's path — there's no authenticated session here for that gate to meaningfully apply to, but
/// a signed-in caller's token (if one happens to be attached) would otherwise still authenticate and
/// trip it. Rate-limited (same IP-partitioned sliding-window mechanism as MigrationController/
/// AuthController, its own "telemetry" policy — see Program.cs) since this is a new unauthenticated
/// write surface.
/// </summary>
[EnableRateLimiting("telemetry")]
[ApiController]
[AllowAnonymous]
[Route("api/telemetry")]
public class TelemetryController : ControllerBase
{
    private readonly TelemetryService _telemetry;

    public TelemetryController(TelemetryService telemetry)
    {
        _telemetry = telemetry;
    }

    [HttpPost("page-load")]
    public async Task<IActionResult> ReportPageLoad(ReportPageLoadRequest request)
    {
        await _telemetry.RecordPageLoadAsync(request.DurationMs);
        return NoContent();
    }
}

File: php-api/src/Controllers/TelemetryController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\TelemetryService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from Controllers/TelemetryController.cs. Anonymous Real User Monitoring beacon — no auth
 * requirement at all (unlike every other controller in this API), since it's a fire-and-forget
 * report from every page load, signed in or not. See SessionValidationMiddleware.php, which
 * explicitly excludes this route's path from the MustChangePassword gate — there's no authenticated
 * session here for that gate to meaningfully apply to. Rate-limited under its own "telemetry" policy
 * (see routes.php) rather than "auth"'s brute-force-tuned limit.
 */
final class TelemetryController extends BaseController
{
    public function reportPageLoad(Request $request, Response $response): Response
    {
        $body = $this->body($request);
        $service = new TelemetryService(Database::connection());
        $service->recordPageLoad($body['durationMs'] ?? null);
        return $this->noContent($response);
    }
}
